src/ripple/protocol/impl/AccountID.cpp

```diff
@@ -25,9 +25,135 @@
 
 namespace ripple {
 
+#ifndef RIPPLED_ACCOUNTID_CACHE_SIZE
+#define RIPPLED_ACCOUNTID_CACHE_SIZE 1000000
+#endif
+
+namespace detail {
+
+/** Caches the base58 representations of AccountIDs */
+class AccountIdCache
```

Comment:
Please ignore the fact that I "requested changes". I can't find a way to undo the request.

Reply:
I believe that we still have the `std::string` new/delete overhead, because `encodeBase58Token` returns a std::string (which is destroyed immediately).

Reply:
Thanks for the review!

>If, instead of storing a char * as the mapped_type, we store the character buffer itself (for ex std::array<char, size_>), we wouldn't need to do the manual memory management of the memory block, and the number of allocations would be the same (with the benefit that we wouldn't have to preallocate a giant memory block at the beginning).

I agree that using `std::array` would be nice to avoid the need for the memory allocation/deallocation but I don't think we can since the size of the buffer is not known at compile time. Unless I am missing something?

I could have used a `std::vector` really, but given that we're only doing the allocation once at start-up and (effectively) never deallocating (the `std::unique_ptr` will go out of scope, if the process shuts down cleanly, but by then we're well into shutting down). I think the "risk" and the programmer overhead of managing the lifetime is basically zero.


>When we don't have a cache hit, we still have the std::string new/delete overhead, because encodeBase58Token returns a std::string (which is destroyed immediately). It would be nice if we had a version that used a provided buffer.

Agreed, hence the [comment](https://github.com/ripple/rippled/pull/4181/files#diff-e5fa0ea64bdeab9635beb8f04eeb2fc076d0fa6df2e133a01209ab3ff33905eaR126-R127) I left to that effect. I think it's worth adding such an API, but I didn't want to do this as part of this commit.


>Now the issue is what should we do when the hash_map is full. Ideally we'd want to replace the element in the array that is the first probe of the hash value. I don't think there is a way to do this with the standard unordered_map

I'm pretty sure that that's what this commit tries to do: it finds the bucket where the new item belongs and replaces the first item with it; if there's no such item (i.e. that bucket is empty) it finds the first item in the next bucket, wrapping around if necessary. It's not _perfect_, in that it still does a `remove` and an `insert`, but I think that's close enough to the functionality you're suggesting, no?

Reply:
Thanks Nik!

> I agree that using std::array would be nice to avoid the need for the memory allocation/deallocation but I don't think we can since the size of the buffer is not known at compile time. Unless I am missing something?

Maybe I'm misunderstanding something. I thought that each `char *` inserted in the map points to a constant size buffer of 40 characters (inside the big block allocated at the start).

Since this map entry is dynamically allocated anyways, I was suggesting to replace the `char *` in the `unordered_map` with the 40 character buffer itself. It would not change the number of allocations, but be simpler and allocate memory as needed instead of all at the beginning.

> It's not perfect, in that it still does a remove and an insert, but I think that's close enough to the functionality you're suggesting, no?

Yes, that part is equivalent. The main thing I was suggesting was that, if we really want to reduce memory allocations to the minimum, we could use a hash map that doesn't allocate a bucket per entry, but one that stores all entries in a big array.

Reply:
This is what I was suggesting (untested)

```
class AccountIdCache
{
private:
    // The total number of items the cache buffer can contain
    std::size_t const count_;

    // The maximum size (in bytes) of an item, along with a null terminator
    // and aligned at 8 bytes
    static constexpr std::size_t const size_ = 40;

    // The offset of the next available block of `size_` contiguous bytes
    std::size_t index_ = 0;

    struct  base58_str
    {
        explicit base58_str(std::string const &val) {
            assert(val.size() < size_);
            std::strcpy(str.data(), val.c_str());
        }
        
        std::array<char, size_> str;
    };

    // The map
    hardened_hash_map<AccountID, base58_str> map_;

    std::mutex mutable mutex_;

    // The total number of requests to format an AccountID
    std::size_t requests_ = 0;

    // The number of times were able to satisfy the request from cache
    std::size_t hits_ = 0;

public:
    AccountIdCache(std::size_t count)
        : count_(count)
        , map_(count_ / 971)
    {
    }

    std::string
    toBase58(AccountID const& id)
    {
        std::lock_guard lock(mutex_);

        ++requests_;

        // We use the map's bucket interface:
        auto const b = map_.bucket(id);

        auto iter = std::find_if(map_.begin(b), map_.end(b), [&id](auto& me) {
            return me.first == id;
        });

        if (iter != map_.end(b))
        {
            ++hits_;
            return iter->second.str.data();
        }

        assert (index_ <= count_);

        if (index_ == count_) {
            // Remove the first item from the first non-empty bucket,
            // starting with this one:
            for (std::size_t i = 0; i != map_.bucket_count();  ++i)
            {
                if (map_.bucket_size((b + i) % map_.bucket_count()) != 0)
                {
                    iter = map_.begin((b + i) % map_.bucket_count());

                    assert(iter != map_.end((b + i) % map_.bucket_count()));

                    map_.erase(iter->first);
                }
            }
        }

        // This copy could be optimized away if we had a "dangerous"
        // interface that accepted the output buffer as an argument.
        auto base58_str =  encodeBase58Token(TokenType::AccountID, id.data(), id.size());
        map_.emplace(id, base58_str);

        return std::move(base58_str);
    }
};
```

Reply:
When `map_.emplace(id, base58_str);` is called, it dynamically allocates a new entry containing the map::value_type and the linked list pointers. If that new entry contains the 40 bytes buffer (instead of a char *), we piggyback on the call to `new` that the hash map does internally, we don't allocate a separate buffer.

Reply:
I learned something new today; thanks! I think your approach would be an improvement over the _existing_ cache and perhaps even the original version of my PR, however, after thinking about your comments more broadly, I decided to adopt a different approach and, effectively, eliminate the `hardened_hash_map` aka `std::unordered_map` altogether. Curious what you think of the changes!

Reply:
If removing the ordered_map  doesn't affect performance significantly, then I fully agree with removing it. In my opinion any added complexity should have significant benefits, otherwise it is not worth it. After adding seelabs new base64 implementation, we can always revisit the cache idea and see if we can come up with something which is a clear improvement.